### PR TITLE
Update mypy version

### LINKED ISSUE #15159: Use a newer version of `mypy==0.971`
## Proposed change
Apply the following patch (as of now) to `requirements/pytorch/test.txt` just like #15158 and fix typing following the raised errors:
```diff
-mypy==0.971
+mypy==0.982
```

### Motivation
To stay up-to-date :)

### Additional context
As we've seen in https://github.com/Lightning-AI/lightning/pull/15158#issuecomment-1281698911, the new mypy 0.982 raises around 40 errors in our codebase.

<details>
<summary>click here for mypy errors</summary>

```
Found 40 errors in 26 files (checked 417 source files)
src/lightning_lite/lite.py:172: error: Unused "type: ignore" comment
src/pytorch_lightning/profilers/pytorch.py:345: error: Incompatible return value type (implicitly returns "None", expected "Union[int, float]")  [return-value]
src/pytorch_lightning/core/saving.py:146: error: Incompatible return value type (implicitly returns "None", expected "Union[LightningModule, LightningDataModule]")  [return-value]
src/pytorch_lightning/trainer/connectors/logger_connector/result.py:228: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/epoch/prediction_epoch_loop.py:50: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/epoch/prediction_epoch_loop.py:70: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:73: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:105: error: Unused "type: ignore" comment
src/pytorch_lightning/utilities/logger.py:91: error: Incompatible return value type (implicitly returns "None", expected "List[Any]")  [return-value]
src/pytorch_lightning/tuner/lr_finder.py:104: error: Incompatible return value type (implicitly returns "None", expected "Callable[[Trainer], None]")  [return-value]
src/pytorch_lightning/loggers/wandb.py:365: error: Unused "type: ignore" comment
src/pytorch_lightning/loggers/tensorboard.py:156: error: Unused "type: ignore" comment
src/pytorch_lightning/loggers/neptune.py:323: error: Unused "type: ignore" comment
src/pytorch_lightning/loggers/neptune.py:353: error: Unused "type: ignore" comment
src/pytorch_lightning/loggers/mlflow.py:155: error: Unused "type: ignore" comment
src/pytorch_lightning/loggers/csv_logs.py:172: error: Unused "type: ignore" comment
src/pytorch_lightning/loggers/comet.py:268: error: Unused "type: ignore" comment
src/pytorch_lightning/core/module.py:602: error: Incompatible return value type (implicitly returns "None", expected "Union[Tensor, Dict[str, Any]]")  [return-value]
src/pytorch_lightning/callbacks/progress/rich_progress.py:380: error: Incompatible return value type (implicitly returns "None", expected "TaskID")  [return-value]
src/pytorch_lightning/loops/fit_loop.py:200: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/fit_loop.py:221: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/fit_loop.py:253: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/optimization/optimizer_loop.py:191: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/optimization/optimizer_loop.py:198: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/optimization/manual_loop.py:101: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/epoch/training_epoch_loop.py:1[59](https://github.com/Lightning-AI/lightning/actions/runs/3269716426/jobs/5377534254#step:6:60): error: Unused "type: ignore" comment
src/pytorch_lightning/loops/epoch/training_epoch_loop.py:174: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/dataloader/prediction_loop.py:84: error: Unused "type: ignore" comment
src/pytorch_lightning/loops/batch/training_batch_loop.py:[61](https://github.com/Lightning-AI/lightning/actions/runs/3269716426/jobs/5377534254#step:6:62): error: Unused "type: ignore" comment
src/pytorch_lightning/loops/batch/training_batch_loop.py:[70](https://github.com/Lightning-AI/lightning/actions/runs/3269716426/jobs/5377534254#step:6:71): error: Unused "type: ignore" comment
src/pytorch_lightning/lite/lite.py:1[79](https://github.com/Lightning-AI/lightning/actions/runs/3269716426/jobs/5377534254#step:6:80): error: Incompatible return value type (implicitly returns "None", expected "Strategy")  [return-value]
src/pytorch_lightning/demos/boring_classes.py:98: error: Unused "type: ignore" comment
src/pytorch_lightning/demos/boring_classes.py:110: error: Unused "type: ignore" comment
src/pytorch_lightning/demos/boring_classes.py:122: error: Unused "type: ignore" comment
src/pytorch_lightning/demos/boring_classes.py:131: error: Unused "type: ignore" comment
src/pytorch_lightning/demos/boring_classes.py:197: error: Unused "type: ignore" comment
src/pytorch_lightning/demos/boring_classes.py:214: error: Unused "type: ignore" comment
src/pytorch_lightning/demos/boring_classes.py:217: error: Unused "type: ignore" comment
```

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by Atharva-Phatak**
## What does this PR do?
This PR updates the mypy version as mentioned in #15159 

Fixes #15159 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
Wondering why dependabot did not find it...

**Comment by otaj** [comment]
> Wondering why dependabot did not find it...

Dependabot found it, but things were breaking, therefore @akihironitta closed it https://github.com/Lightning-AI/lightning/pull/15158#issuecomment-1281698938

**Review by Borda — APPROVED** [review]

**Comment by awaelchli** [comment]
@Atharva-Phatak the mypy check needs to be resolved (Unused "type: ignore" comment). Could you look into it?

**Comment by Atharva-Phatak** [comment]
@awaelchli I thought they would be resolved with an updated version of mypy. I will remove these comments.

**Comment by otaj** [comment]
> @awaelchli I thought they would be resolved with an updated version of mypy. I will remove these comments.

CI is using the version specified in the requirements - therefore mypy errors you are seeing in CI is originating from updated mypy version already. Let us know if you need any help with it :zap:

**Comment by Atharva-Phatak** [comment]
@awaelchli Could you help me out here ? I am understanding what is going wrong here. Sorry :)

File: src/pytorch_lightning/utilities/parameter_tying.py [hunk 1]
```diff
@@ -44,8 +41,7 @@ def _find_shared_parameters(module: nn.Module, tied_parameters: Optional[Dict] =
             continue
         submodule_prefix = prefix + ("." if prefix else "") + name
         _find_shared_parameters(m, tied_parameters, submodule_prefix)
-    if first_call:
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/parameter_tying.py:47]: I don't think this was required, other than for the improvement in efficiency of not returning anything for the inner recursive calls

File: src/pytorch_lightning/utilities/logger.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def _flatten_dict(params: Dict[Any, Any], delimiter: str = "/") -> Dict[str, Any
 
     def _dict_generator(
         input_dict: Any, prefixes: List[Optional[str]] = None
-    ) -> Generator[Any, Optional[List[str]], List[Any]]:
+    ) -> Generator[Any, Optional[List[str]], List[Any]]:  # FIXME
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/logger.py:93]: FIXME: Not sure about this one yet...

**Review by carmocca — COMMENTED** [review]
Fixed all errors but one

File: src/pytorch_lightning/utilities/logger.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/utilities/logger.py:93]: Maybe ignore it? 😃

**Review by awaelchli — APPROVED** [review]

**Comment by Atharva-Phatak** [comment]
Still, the mypy tests are failing 🤔

File: src/pytorch_lightning/utilities/logger.py (hunk 2 above)
Comment (carmocca) [reply src/pytorch_lightning/utilities/logger.py]: I instead simplified the implementation so typing is not ambiguous

**Comment by carmocca** [comment]
This should be good now, but I'd wait until 1.8 is released to merge it just in case we it breaks something

**Review by carmocca — APPROVED** [review]
